data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-transform/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/transformer/type/to/JsonObjectToTransferCompletionMessageTransformer.java

```diff
@@ -0,0 +1,57 @@
+/*
+ *  Copyright (c) 2023 Fraunhofer Institute for Software and Systems Engineering
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Fraunhofer Institute for Software and Systems Engineering - initial API and implementation
+ *
+ */
+
+package org.eclipse.edc.protocol.dsp.transferprocess.transformer.type.to;
+
+import jakarta.json.JsonObject;
+import org.eclipse.edc.connector.transfer.spi.types.protocol.TransferCompletionMessage;
+import org.eclipse.edc.jsonld.transformer.AbstractJsonLdTransformer;
+import org.eclipse.edc.protocol.dsp.spi.types.HttpMessageProtocol;
+import org.eclipse.edc.transform.spi.TransformerContext;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import static java.lang.String.format;
+import static org.eclipse.edc.protocol.dsp.transferprocess.transformer.DspCatalogPropertyAndTypeNames.DSPACE_CALLBACKADDRESS_TYPE;
+import static org.eclipse.edc.protocol.dsp.transferprocess.transformer.DspCatalogPropertyAndTypeNames.DSPACE_PROCESSID_TYPE;
+import static org.eclipse.edc.protocol.dsp.transferprocess.transformer.DspCatalogPropertyAndTypeNames.DSPACE_TRANSFER_COMPLETION_TYPE;
+
+public class JsonObjectToTransferCompletionMessageTransformer extends AbstractJsonLdTransformer<JsonObject, TransferCompletionMessage> {
+
+    public JsonObjectToTransferCompletionMessageTransformer() {
+        super(JsonObject.class, TransferCompletionMessage.class);
+    }
+
+    @Override
+    public @Nullable TransferCompletionMessage transform(@NotNull JsonObject jsonObject, @NotNull TransformerContext context) {
+        var type = nodeType(jsonObject, context);
+
+
+        if (DSPACE_TRANSFER_COMPLETION_TYPE.equals(type)) {
```

Comment:
Many of these transformers have the same implementation pattern: they check the type and register a problem if it doesn't match. Aren't these transformers dispatched to by type? If so, these should be refactored to avoid this and perform a one-line assertion at the start of the method. Currently, these checks are cluttering all of the implementations.

Reply:
Let's take `JsonObjectToTransferCompletionMessageTransformer` as an example.

A DSP JSON-LD message is deserialized off the wire by Jersey and passed to the DSP controller. The controller needs to validate the incoming message type is correct (or delegate validation). If the message type is incorrect, it should respond with a BAD_REQUEST since the client sent an invalid request according to the DSP specifications. The transformers should not perform this type of error handling as it will result in awkward code and difficulty propagating sufficient error information back to the controller so that it can send the appropriate response to the client. 
 
In other words, the transformers may encounter other problems, but each parent (the controller or a recursively invoked transformer) should ensure valid type dispatching. This will greatly simplify the transformer implementations.

Reply:
I see your point, but can we do it in an additional Issue and refactor this later.

Because this implementation pattern is already used in the dsp-catalog-transform modul and dsp-transform modul.

And do it in the same way for all the transform modules in the dsp?

Reply:
We should fix it now so as not to introduce more technical debt. Those other examples slipped through and they also need to be fixed.

Reply:
Removed check, weill be done in API